src/ZODB/Connection.py

```diff
@@ -1083,6 +1042,7 @@ def _resetCache(self):
             self._reader._cache = cache
 
     def _release_resources(self):
+        assert not self.opened
```

Comment:
I guess this is a place where our styles clash:  I don't like `assert` in non-test code.